pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ReferenceResolver is an interface for resolving cross resource attribute references.
+// It should be implemented by strcuts that are used for custom resources feilds.
+// See #874 for more information
+type ReferenceResolver interface {
+	// Resolve retrieve and builds the referenced attribute
+	Resolve(reader client.Reader) (value interface{}, err error)
```

Comment:
Could `value` be `string` rather than `interface{}`? I prefer to avoid the empty interface where possible, and I can't think of many (any?) cases where a reference would be resolved to an integer, byte array, struct, etc.

Reply:
I can change it to string for now. But I want to be vary of limitations that this will imply in the future.